Boot test AMIs in AWS

achilleas-k opened the pull request:
When building a test AMI, upload it to AWS and boot it there, running the same smoke test as we do for the qcow2.
The local testing of the AMI is still run as before.

Added two subclasses to the VM class:
- QEMU: is the old code that initialises and boots an image in qemu.
- AWS: new code that launches (and terminates) an AMI.

mvo5 commented on test/vm.py:
```diff
@@ -3,13 +3,42 @@
 
 
-from paramiko.client import AutoAddPolicy, SSHClient
+class VM:
```
(nitpick) we could use a abstract base class here but maybe I'm overthinking it

achilleas-k replied:
Done.

mvo5 commented on plans/all.fmf:
```diff
@@ -13,6 +13,7 @@ prepare:
     - python3-boto3
     - python3-flake8
     - python3-paramiko
+    - python3-boto3
```
(nipick^3) this change shold probably be part of commit abde26512f9d3238214f913458a2e3bd607f4283

achilleas-k replied:
(it's actually a duplicate)

achilleas-k replied:
Done.

mvo5 commented on test/vm.py:
```diff
@@ -103,3 +106,114 @@ def run(self, cmd, user, password):
+            self._set_ssh_security_group()
+        sec_group_ids = [self._ec2_security_group.id]
+        try:
+            print(f"Creating ec2 instance from {self._ami_id}")
```
(nitpick) if we use `self._log()` here we will be pepared for the time when we have better logging for this :) (and in the other places that now use print())

achilleas-k replied:
Not a nitpick.  You're absolutely right and we should.

achilleas-k replied:
Done.

mvo5 commented on test/vm.py:
```diff
@@ -103,3 +106,114 @@ def run(self, cmd, user, password):
+            print("No security group defined. Skipping deletion.")
+
+    def run(self, cmd, user, password):
+        if not self._ec2_instance:
```
if we make this `self.running()` and rename `self._ec_instance` to something like `_ip` we could share the run() code unchanged between QEMU/ec2 and move it into the VM base class, wdyt?

achilleas-k replied:
I think I was expecting the `run()` function to differ more when I started and didn't revisit it after everything was done.  It makes total sense  to share it, I agree.

achilleas-k replied:
Done.

mvo5 commented on test/testutil.py:
```diff
@@ -77,3 +80,27 @@ def can_start_rootful_containers():
+        creds_file.write(f"aws_secret_access_key = {secret_key}\n")
+
+
+def delete_ami(ami_id):
```
(super nitpick) maybe `deregister_ami()` to follow upstream conventions?

achilleas-k replied:
Done.

mvo5 commented on test/test_smoke.py:
```diff
@@ -166,6 +167,21 @@ def test_image_boots(image_type):
+@pytest.mark.skipif(platform.system() != "Linux", reason="boot test only runs on linux right now")
+@pytest.mark.parametrize("image_type", CLOUD_IMAGE_TYPES, indirect=["image_type"])
+def test_image_boots_cloud(image_type):
+    if image_type.img_type == "ami":
```
I wonder if it would make sense to just call this `test_image_boots_ec2` or `test_ami_boots_in_cloud` and hard code `@pytest.mark.parametrize("image_type", ["ami"], indirect=["image_type"])` - mostly to the `if image_type.img_type == ` checks, if all tests will look slightly different anyway we can as well make them different ?

achilleas-k replied:
Yeah, fair.  It'll just end up being a big `switch` (or `if/else`s) in the end with no commonalities.
Also, we can probably enable it for macOS too (not that I'm looking at the pytest.marks again.

achilleas-k replied:
Done.

mvo5 commented on test/testutil.py:
```diff
@@ -75,3 +80,27 @@ def can_start_rootful_containers():
+    key_id = os.environ.get("AWS_ACCESS_KEY_ID")
+    secret_key = os.environ.get("AWS_SECRET_ACCESS_KEY")
+    if not key_id or not secret_key:
+        print("WARNING: aws credentials not available")
```
Should this be an error (i.e. raise here)?

achilleas-k replied:
I was thinking whether this should be an error or not and thinking about it again (and seeing how it actually failed because the secrets aren't set), I can't think of a good reason for it being just a warning.

achilleas-k replied:
Done.

mvo5 commented on test/vm.py:
```diff
@@ -1,15 +1,47 @@
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
```
I think it would be nice if we could have a shared `run()` here as the ssh bits are a bit annoying to get right we can avoid some copy/paste this way :) We most likely need a common `running(), _ssh_port, _ip` between the classes.

achilleas-k replied:
Done.

mvo5 reviewed (commented):
Thanks for the PR! This is super nice and it's great to see how well this all seem to fit. I have a bunch of nipticks/ideas inline but all can be followups if you prefer that :)

mvo5 commented on test/test_smoke.py:
```diff
@@ -95,22 +100,48 @@ def image_type_fixture(tmpdir_factory, build_container, request):
+    upload_args = []
+    creds_args = []
+    with tempfile.TemporaryDirectory() as tempdir:
+        if image_type == "ami":
```
we probably need a check there that we actually have AWS credentials, when running this on a system without we should skip this part of the test (but not the ami generation as this is useful with qemu still)

achilleas-k replied:
Not sure about this.  Currently it will fail, which can be annoying if you want to test the AMI in a scenario where you don't have an AWS account but intend to boot it with QEMU.  But if we skip the test instead, then a config mixup in CI will hide the issue by ending up green.

achilleas-k commented:
https://github.com/osbuild/bootc-image-builder/pull/93#discussion_r1446217548

Added a new pytest flag, `--force-aws-upload`.  Setting this flag makes the AMI build fail if the credentials aren't set.  Without it, the test is skipped.
Apply the following patch and run the following to quickly see how it works:
```diff
diff --git a/test/test_smoke.py b/test/test_smoke.py
index 9703240..eb6ddc2 100644
--- a/test/test_smoke.py
+++ b/test/test_smoke.py
@@ -119,6 +119,8 @@ def image_type_fixture(tmpdir_factory, build_container, request, force_aws_uploa
                 # upload forced but credentials aren't set
                 raise RuntimeError("AWS credentials not available (upload forced)")
 
+            return
+
         # run container to deploy an image into a bootable disk and upload to a cloud service if applicable
         subprocess.check_call([
             "podman", "run", "--rm",
```

```
$ sudo pytest -k test_ami_boots -v --force-aws-upload

================================================= test session starts =================================================
platform linux -- Python 3.11.6, pytest-7.4.4, pluggy-1.3.0 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /home/achilleas/projects/osbuild/bootc-image-builder
plugins: forked-1.6.0, xdist-3.5.0, anyio-4.2.0
collected 14 items / 13 deselected / 1 selected

test/test_smoke.py::test_ami_boots_in_aws[ami] ERROR                                                            [100%]

======================================================= ERRORS ========================================================
____________________________________ ERROR at setup of test_ami_boots_in_aws[ami] _____________________________________

<snip>

test/test_smoke.py:120: RuntimeError
=============================================== short test summary info ===============================================
ERROR test/test_smoke.py::test_ami_boots_in_aws[ami] - RuntimeError: AWS credentials not available (upload forced)
=========================================== 13 deselected, 1 error in 1.89s ===========================================

$ sudo pytest -k test_ami_boots -v
================================================= test session starts =================================================
platform linux -- Python 3.11.6, pytest-7.4.4, pluggy-1.3.0 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /home/achilleas/projects/osbuild/bootc-image-builder
plugins: forked-1.6.0, xdist-3.5.0, anyio-4.2.0
collected 14 items / 13 deselected / 1 selected

test/test_smoke.py::test_ami_boots_in_aws[ami] SKIPPED (AWS credentials not available (upload not forced))      [100%]

========================================== 1 skipped, 13 deselected in 1.79s ==========================================
```

mvo5 reviewed (approved):
I love it!

achilleas-k commented:
> I love it!

We'll anyway need to rebase once #95 gets merged.  I was hoping that would be merged while I was working on the review comments here, otherwise I would have added it in this one.

achilleas-k commented:
Seems the GH secrets aren't being picked up:
* Added them as [envs to the workflow](https://github.com/osbuild/bootc-image-builder/blob/51a3d7f5895fc4da49b5b9ec4aa152a3836298bc/.github/workflows/tests.yml#L73-L81)
* Added them as [org secrets](https://github.com/osbuild/bootc-image-builder/settings/secrets/actions)
* But they're not being picked up in the run: [log: step setup](https://github.com/osbuild/bootc-image-builder/actions/runs/7464604252/job/20311939290?pr=93#step:6:5) and [log: test failure](https://github.com/osbuild/bootc-image-builder/actions/runs/7464604252/job/20311939290?pr=93#step:6:862)

(The value should appear as `***` on lines 5 and 6 in the log, like [here](https://github.com/osbuild/bootc-image-builder/actions/runs/7456549403/job/20287371887#step:3:6))

Feel like I'm missing something obvious.

mvo5 commented:
> Seems the GH secrets aren't being picked up:
> 
>     * Added them as [envs to the workflow](https://github.com/osbuild/bootc-image-builder/blob/51a3d7f5895fc4da49b5b9ec4aa152a3836298bc/.github/workflows/tests.yml#L73-L81)
> 
>     * Added them as [org secrets](https://github.com/osbuild/bootc-image-builder/settings/secrets/actions)
> 
>     * But they're not being picked up in the run: [log: step setup](https://github.com/osbuild/bootc-image-builder/actions/runs/7464604252/job/20311939290?pr=93#step:6:5) and [log: test failure](https://github.com/osbuild/bootc-image-builder/actions/runs/7464604252/job/20311939290?pr=93#step:6:862)
> 
> 
> (The value should appear as `***` on lines 5 and 6 in the log, like [here](https://github.com/osbuild/bootc-image-builder/actions/runs/7456549403/job/20287371887#step:3:6))
> 
> Feel like I'm missing something obvious.

I (strongly) suspect that this is because GH secrets are not available in repo forks. The easiest fix might be to just move this test into `testingfarm.yml` that already has the (sightly involved) setup to workaround this.

achilleas-k commented:
Rebased and fixed conflict.  Still needs #97.

mvo5 reviewed (dismissed):
Thank you!
